pkg/ipcache/ipcache.go

```diff
@@ -759,11 +761,11 @@ func (ipc *IPCache) GetNamedPorts() (npm types.NamedPortMultiMap) {
 
 // DeleteOnMetadataMatch removes the provided IP to security identity mapping from the IPCache
 // if the metadata cache holds the same "owner" metadata as the triggering pod event.
-func (ipc *IPCache) DeleteOnMetadataMatch(IP string, source source.Source, namespace, name string) (namedPortsChanged bool) {
+func (ipc *IPCache) DeleteOnMetadataMatch(IP string, source source.Source, namespace, name, uid string) (namedPortsChanged bool) {
 	ipc.mutex.Lock()
 	defer ipc.mutex.Unlock()
 	k8sMeta := ipc.getK8sMetadata(IP)
-	if k8sMeta != nil && k8sMeta.Namespace == namespace && k8sMeta.PodName == name {
+	if k8sMeta != nil && k8sMeta.Namespace == namespace && k8sMeta.PodName == name && k8sMeta.UID == uid {
```

Comment:
Isn't `k8sMeta.UID` going to be always empty for restored endpoints after upgrade :thinking: ?

Reply:
It will be and so this check will be true, so it shouldn't be an issue. When a new pod is created while running the new code, then the UID will be non-empty. Either the UID is set or it's empty, in other words, there shouldn't be a situation where it's mixed. So I don't see an issue, do you have something in mind?

Reply:
One approach we can take here is to update the endpointsynchronizer `sync-to-k8s-ciliumendpoint` controller to also set the pod UID in the `Endpoint` object similar to what's being done already for CEP UID. However, there's no guarantee that this controller runs before [the ipache code above](https://github.com/cilium/cilium/blob/dbb579ada86ebcde8bf41b10a55b23528099d4f8/pkg/ipcache/ipcache.go#L332) is invoked. I'm not sure of another way around this than to skip the UID check if the UID is empty in [that ipcache code](https://github.com/cilium/cilium/blob/dbb579ada86ebcde8bf41b10a55b23528099d4f8/pkg/ipcache/ipcache.go#L332). What do you think?

Reply:
https://github.com/cilium/cilium/pull/31605#issuecomment-2097483504